engine/src/main/java/org/pentaho/di/trans/steps/excelinput/ods/OdfWorkbook.java

```diff
@@ -0,0 +1,124 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2018 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.trans.steps.excelinput.ods;
+
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.odftoolkit.odfdom.doc.OdfDocument;
+import org.odftoolkit.odfdom.doc.OdfSpreadsheetDocument;
+import org.odftoolkit.odfdom.doc.table.OdfTable;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.spreadsheet.KSheet;
+import org.pentaho.di.core.spreadsheet.KWorkbook;
+import org.pentaho.di.core.vfs.KettleVFS;
+
+public class OdfWorkbook implements KWorkbook {
+
+  private String filename;
+  private String encoding;
+  private OdfDocument document;
+  private Map<String, OdfSheet> openSheetsMap = new HashMap<String, OdfSheet>();
+
+  public OdfWorkbook( String filename, String encoding ) throws KettleException {
+    this.filename = filename;
+    this.encoding = encoding;
+
+    try {
+      document = OdfSpreadsheetDocument.loadDocument( KettleVFS.getInputStream( filename ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when OdfSpreadsheetDocument.loadDocument throws an exception. Ensure the object returned by *getInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).